### PR TITLE
EdgeFluxRegister for MHD

### LINKED ISSUE #2512: 'flux registers' for edge-centered EMFs?
### Discussed in https://github.com/AMReX-Codes/amrex/discussions/2511

<div type='discussions-op-text'>

<sup>Originally posted by **BenWibking** December  2, 2021</sup>
Hi,

We are planning to implement face-centered MHD in Quokka. Is there a 'flux register' to do the correction at coarse-fine interfaces of edge-centered EMFs in AMReX? In Chombo, this can be accomplished with [`LevelFluxRegisterEdge`](http://colella.lbl.gov/Manuals/CHOMBO-RELEASE-3.3/classLevelFluxRegisterEdge.html), which has a similar API to `YAFluxRegister`.

Example:
Screen Shot 2021-12-03 at 2 14 21 PM

There are a number of corner cases (see panels a-d copied below). (In [Athena++](https://arxiv.org/pdf/2005.06651.pdf), this step is also done at all shared edges between grids on the same resolution level in order to avoid roundoff errors, but it's not currently clear whether this will be necessary.)

Screen Shot 2021-12-03 at 2 15 11 PM</div>

### PR DESCRIPTION
**Pull request by WeiqunZhang**
This implements EdgeFluxRegister for refluxing in MHD. The equation here is ${\partial B} / {\partial t} + \nabla \times E = 0,$ where B is on faces and E is on edges.

Close #2512

## Checklist

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [x] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] include documentation in the code and/or rst files, if appropriate

### CODE REVIEW

**Comment by WeiqunZhang** [comment]
I have done some testing using https://github.com/WeiqunZhang/amrex-devtests/blob/main/edgefluxregister/

File: Src/Boundary/AMReX_EdgeFluxRegister.H [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+ *
+ * If the equation has an extra factor due to the choice of units, the
+ * factor can be absorbed into dt. If we have `v x B` instead of E, the sign
+ * can also been absorbed into dt. Note that whatever the choice of sign is,
+ * the dt arguments passed to CrseAdd and FineAdd should have the same sign.
+ *
+ * We try to keep the interface simple by not providing overloads that
+ * specify the component index. If the user's data does not start with
+ * component 0, it can be worked around by creating alias FArrayBox and
+ * MulitiFab.
```
Comment (eebasso) [inline Src/Boundary/AMReX_EdgeFluxRegister.H]: "MulitiFab" -> "MultiFab"

File: Src/Boundary/AMReX_EdgeFluxRegister.H [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    void average_down_faces (const Array<const MF*,AMREX_SPACEDIM>& fine,
+                             const Array<MF*,AMREX_SPACEDIM>& crse,
+                             const IntVect& ratio, const Geometry& crse_geom)
+ * \endverbatim
+ *
+ * Note that both CrseAdd and FineAdd are async for GPU builds. That means
+ * it's the user's responsibility to keep the FArrayBox arguments alive or
+ * call Gpu::streamSynchronize() when necessary.
+ *
+ * Because stagger grids are used, tiling could be very confusing. To avoid
```
Comment (eebasso) [inline Src/Boundary/AMReX_EdgeFluxRegister.H]: "staggered"

**Comment by eebasso** [comment]
Beginner question: why in 2D do we keep $B_x$, $B_y$, and $E_z$? I would have expected $E_x$, $E_y$, and $B_z$ if the particles are restricted to moving in the x-y plane.

**Comment by WeiqunZhang** [comment]
This is for MHD, where $E = - v \times B$. When both v and B are in the x-y plane, E is in z-direction only.

For ideal MHD, we have $j = \sigma (E + v \times B)$ and $\sigma = \infty$. Therefore, $E + v \times B = 0$.

**Review by atmyers — APPROVED** [review]

**Comment by WeiqunZhang** [comment]
@BenWibking Let us know if you notice any issues. Thanks!

**Pull request merged**